File: 2024/day19/node/solution.js
import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const input = readFileSync(join(__dirname, '..', 'input.txt'), 'utf-8').trim();

// Parse input
const parts = input.split('\n\n');
const patterns = parts[0].split(',').map(p => p.trim());
const designs = parts[1].trim().split('\n');

// Part 1
function canForm(design) {
  const memo = new Map();

  function dp(pos) {
    if (pos === design.length) return true;
    if (memo.has(pos)) return memo.get(pos);

    for (const pattern of patterns) {
      const plen = pattern.length;
      if (design.slice(pos, pos + plen) === pattern) {
        if (dp(pos + plen)) {
          memo.set(pos, true);
          return true;
        }
      }
    }
    memo.set(pos, false);
    return false;
  }

  return dp(0);
}

function part1() {
  return designs.filter(d => canForm(d)).length;
}

// Part 2
function countWays(design) {
  const memo = new Map();

  function dp(pos) {
    if (pos === design.length) return 1n;
    if (memo.has(pos)) return memo.get(pos);

    let total = 0n;
    for (const pattern of patterns) {
      const plen = pattern.length;
      if (design.slice(pos, pos + plen) === pattern) {
        total += dp(pos + plen);
      }
    }
    memo.set(pos, total);
    return total;
  }

  return dp(0);
}

function part2() {
  return designs.reduce((sum, d) => sum + countWays(d), 0n);
}

console.log('Part 1:', part1());
console.log('Part 2:', part2());

File: 2024/day19/ruby/solution.rb
#!/usr/bin/env ruby

input_path = File.join(File.dirname(__FILE__), "..", "input.txt")
input_text = File.read(input_path).strip

# Parse input
parts = input_text.split("\n\n")
patterns = parts[0].split(",").map(&:strip)
designs = parts[1].strip.split("\n")

def can_form(design, patterns)
  memo = {}

  dp = lambda do |pos|
    return true if pos == design.length
    return memo[pos] if memo.key?(pos)

    result = patterns.any? do |pattern|
      plen = pattern.length
      if design[pos, plen] == pattern
        dp.call(pos + plen)
      else
        false
      end
    end

    memo[pos] = result
  end

  dp.call(0)
end

def count_ways(design, patterns)
  memo = {}

  dp = lambda do |pos|
    return 1 if pos == design.length
    return memo[pos] if memo.key?(pos)

    total = 0
    patterns.each do |pattern|
      plen = pattern.length
      if design[pos, plen] == pattern
        total += dp.call(pos + plen)
      end
    end

    memo[pos] = total
  end

  dp.call(0)
end

def part1(designs, patterns)
  designs.count { |d| can_form(d, patterns) }
end

def part2(designs, patterns)
  designs.sum { |d| count_ways(d, patterns) }
end

puts "Part 1: #{part1(designs, patterns)}"
puts "Part 2: #{part2(designs, patterns)}"
